pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+                                 chain. This can be used to differentiate a same
+                                 kernel appearing multiple times in a loop chain.
+        """
+        # Protect against re-initialization when retrieved from cache
+        if self._initialized:
+            return
+
+        asts = fused_ast
+        if not asts:
+            # If kernels' need be concatenated, discard duplicates
+            kernels = dict(zip([k.cache_key[1:] for k in kernels], kernels)).values()
+            asts = [k._ast for k in kernels]
+        kernels = as_tuple(kernels, (Kernel, host.Kernel, base.Kernel))
+
+        Kernel._globalcount += 1
+        self._kernels = kernels
+        self._name = "_".join([k.name for k in kernels])
+        self._opts = dict(flatten([k._opts.items() for k in kernels]))
+        self._applied_blas = any(k._applied_blas for k in kernels)
+        self._include_dirs = list(set(flatten([k._include_dirs for k in kernels])))
+        self._headers = list(set(flatten([k._headers for k in kernels])))
+        self._user_code = "\n".join(list(set([k._user_code for k in kernels])))
+
+        # Code generation is delayed until actually needed
+        self._ast = asts
+        self._code = None
+
+        self._initialized = True
+
+    def __iter__(self):
+        for k in self._kernels:
+            yield k
+
+    def __str__(self):
+        return "OP2 FusionKernel: %s" % self._name
+
+
+# Parallel loop API
+
+class JITModule(host.JITModule):
+
+    _cppargs = []
+    _libraries = []
+    _extension = 'cpp'
+
+    _wrapper = """
+extern "C" void %(wrapper_name)s(%(executor_arg)s,
+                      %(ssinds_arg)s
+                      %(wrapper_args)s
+                      %(const_args)s);
+void %(wrapper_name)s(%(executor_arg)s,
+                      %(ssinds_arg)s
+                      %(wrapper_args)s
+                      %(const_args)s) {
+  %(user_code)s
+  %(wrapper_decs)s;
+  %(const_inits)s;
+
+  %(executor_code)s;
+}
+"""
+    _kernel_wrapper = """
+%(interm_globals_decl)s;
+%(interm_globals_init)s;
+%(vec_decs)s;
+%(args_binding)s;
+%(tile_init)s;
+for (int n = %(tile_start)s; n < %(tile_end)s; n++) {
+  int i = %(index_expr)s;
+  %(vec_inits)s;
+  i = %(tile_iter)s[%(index_expr)s];
+  %(buffer_decl)s;
+  %(buffer_gather)s
+  %(kernel_name)s(%(kernel_args)s);
+  %(layout_decl)s;
+  %(layout_loop)s
+      %(layout_assign)s;
+  %(layout_loop_close)s
+  i = %(index_expr)s;
+  %(itset_loop_body)s;
+}
+%(interm_globals_writeback)s;
+"""
+
+    @classmethod
+    def _cache_key(cls, kernel, it_space, *args, **kwargs):
+        key = (hash(kwargs['executor']),)
+        all_args = kwargs['all_args']
+        for kernel_i, it_space_i, args_i in zip(kernel, it_space, all_args):
+            key += super(JITModule, cls)._cache_key(kernel_i, it_space_i, *args_i)
+        return key
+
+    def __init__(self, kernel, it_space, *args, **kwargs):
+        if self._initialized:
```

Comment:
If you pull an object out of the cache in `__new__` the `__init__` method is always called, this way you don't reinit objects coming out the cache.